### PR TITLE
[NETBEANS-4443] PHP 8.0 Support: Named Arguments

### PR DESCRIPTION
**Pull request by junichi11**
https://issues.apache.org/jira/browse/NETBEANS-4443

https://wiki.php.net/rfc/named_params

#### Part 1

- Fix the parser
- Add the `NamedArgument`
- Fix the `PHP80UnhandledError`
- Add unit tests for the parser
- Add `IncorrectNamedArgumentsHintError` to handle incorrect usage

##### PHP 8.0

nb-php80-named-arguments-syntax

##### PHP 7.4

nb-php80-named-arguments-syntax-php74

##### Incorrect Named Arguments
Duplicate names:

nb-php80-incorrect-named-arguments-duplicate-names

nb-php80-incorrect-named-arguments-duplicate-names-in-attribute-arg-list

Positional arguments after named arguments

nb-php80-incorrect-named-arguments-positional-after-named

nb-php80-incorrect-named-arguments-positional-after-named-in-attribute-arg-list

Combine named arguments and argument unpacking

nb-php80-incorrect-named-arguments-named-and-unpack

#### Part 2

- Fix the formatting feature (Don't add spaces before ":")
- Add unit tests for the formatting feature

Example code:
```php
array_fill(   start_index  : 0, count   : 100, value:50);
array_fill(
    value:50,    
   count   : 100, 
   start_index     :0
  );

class ParentClass {

    public function __construct(
            public int $foo,
            public string $bar = "",
    ) {
    }

}

class ExampleClass extends ParentClass {

public function __construct(
        public string $test1,
        public ?int $test2 = null,
) {
    parent::__construct(foo: $test1);
}

public function test(int $test1, string|int $test2 = "") {}

public static function staticTest(ExampleClass $exampleClass, int $test) {}
}

$instance = new ExampleClass($test1, test2: $test2);
$instance->test(test1:$test2);
ExampleClass::staticTest(  exampleClass  : $instance, test  : $test2);
ExampleClass::staticTest(
        exampleClass  : $instance,
        test  : $test2,
);
$array_fill = "array_fill";
$array_fill(   start_index   : 0, count : 100, value            : 50);
$array_fill(
        start_index   : 0,
        count : 100,
        value            : 50
    );

test(test(1, 2, 3), test3:test(1, 2, 3), test2   : test(1, test2   : 2, test3   : [1, 2, 3]));
test(
        test(1, 2, 3),
        test3:test(1, 2, 3),
        test2   : test(1, test2   : 2, test3   : [1, 2, 3])
        );

reservedKeyWord(   array         : []);
reservedKeyWord(
        array         : []
  );

test(  1,   2,   test3    :    $test1   === 0 ? 0:1);
test(
1,
2,
test3    :    $test1   === 0 ? 0:1
);

```

Formatted:

```php
array_fill(start_index: 0, count: 100, value: 50);
array_fill(
        value: 50,
        count: 100,
        start_index: 0
);

class ParentClass {

    public function __construct(
            public int $foo,
            public string $bar = "",
    ) {
        
    }

}

class ExampleClass extends ParentClass {

    public function __construct(
            public string $test1,
            public ?int $test2 = null,
    ) {
        parent::__construct(foo: $test1);
    }

    public function test(int $test1, string|int $test2 = "") {
        
    }

    public static function staticTest(ExampleClass $exampleClass, int $test) {
        
    }

}

$instance = new ExampleClass($test1, test2: $test2);
$instance->test(test1: $test2);
ExampleClass::staticTest(exampleClass: $instance, test: $test2);
ExampleClass::staticTest(
        exampleClass: $instance,
        test: $test2,
);
$array_fill = "array_fill";
$array_fill(start_index: 0, count: 100, value: 50);
$array_fill(
        start_index: 0,
        count: 100,
        value: 50
);

test(test(1, 2, 3), test3: test(1, 2, 3), test2: test(1, test2: 2, test3: [1, 2, 3]));
test(
        test(1, 2, 3),
        test3: test(1, 2, 3),
        test2: test(1, test2: 2, test3: [1, 2, 3])
);

reservedKeyWord(array: []);
reservedKeyWord(
        array: []
);

test(1, 2, test3: $test1 === 0 ? 0 : 1);
test(
        1,
        2,
        test3: $test1 === 0 ? 0 : 1
);
```

### CODE REVIEW

**Comment by junichi11** [comment]
Note: other features are not fixed yet (e.g. Code completion). I'll fix them later.

**Comment by junichi11** [comment]
@neilcsmith-net @geertjanw 
I would like to include this in 12.3 if possible. Is it OK?

File: php/php.editor/tools/ASTPHP5Parser.cup [hunk 1]
```diff
@@ -2535,74 +2537,79 @@ non_empty_function_call_parameter_list:paramsList possible_comma
...
 {:
-    List paramsList = new LinkedList();
-    Expression var_ref = new Reference(startleft, varright, var);
-    paramsList.add(var_ref);
-    RESULT = paramsList;
+    RESULT = var;
 :}
 
-| error:expr
+| T_ELLIPSIS:e argument_expr:var
```
Comment (junichi11) [inline php/php.editor/tools/ASTPHP5Parser.cup:2560]: If we use `is_variadic`, it conflicts with `function_call_parameter_list`.

**Comment by junichi11** [comment]
@tmysik Just in case, let's wait for a reply from @neilcsmith-net @geertjanw .

**Comment by neilcsmith-net** [comment]
@junichi11 master isn't branched yet, so go for it! :+1:

**Comment by junichi11** [comment]
@neilcsmith-net Thank you for your reply :)

File: php/php.editor/src/org/netbeans/modules/php/editor/verification/IncorrectNamedArgumentsHintError.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    @Override
+    @NbBundle.Messages("IncorrectNamedArgumentsHintError.displayName=Incorrect Named Arguments")
+    public String getDisplayName() {
+        return Bundle.IncorrectConstructorPropertyPromotionHintError_displayName();
+    }
+
+    @Override
+    @NbBundle.Messages({
+        "# {0} - name",
+        "IncorrectNamedArguments.desc.duplicate.name=Duplicate name: \"{0}\" already exists.",
```
Comment (KacerCZ) [inline php/php.editor/src/org/netbeans/modules/php/editor/verification/IncorrectNamedArgumentsHintError.java:59]: Personally I'd use "Duplicate argument name:" to be clear it's related to argument if there are more hints for given line.

File: php/php.editor/src/org/netbeans/modules/php/editor/verification/IncorrectNamedArgumentsHintError.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+                description,
+                fileObject,
+                new OffsetRange(node.getStartOffset(), node.getEndOffset()),
+                fixes,
+                500
+        ));
+    }
+
+    //~ Inner classes
+    private static final class CheckVisitor extends DefaultVisitor {
```
Comment (KacerCZ) [inline php/php.editor/src/org/netbeans/modules/php/editor/verification/IncorrectNamedArgumentsHintError.java:114]: There is missing `visit(ClassInstanceCreation node)`, so this code is not marked with hint (duplicate arguments):
```php
$date = new \DateTime(datetime: 'now', datetime: 'now');
`

Comment (junichi11) [reply php/php.editor/src/org/netbeans/modules/php/editor/verification/IncorrectNamedArgumentsHintError.java:114]: Will fix. Thanks.

**Comment by KacerCZ** [comment]
@junichi11 Thanks for implementing this feature.
I tested it and it works as expected.

**Comment by KacerCZ** [comment]
According to RFC name of argument can be also reserved keyword.

In following code
```php
function searchInArray(array $array) {}
searchArray(array: []);
```
NetBeans highlights `array` on second line even if it is not keyword.
Similar case is also for other keywords (`instanceof`, `string`, ...)

This is not big problem and I don't know if it can be easily fixed.

**Comment by junichi11** [comment]
> According to RFC name of argument can be also reserved keyword.

Yes, I already noticed that. As you wrote, it's not a big problem. Maybe, it's not easy. I'm going to try to fix it later.

**Review by KacerCZ — APPROVED** [review]

File: php/php.editor/tools/ASTPHP5Parser.cup (hunk 1 above)
Comment (tmysik) [reply php/php.editor/tools/ASTPHP5Parser.cup:2560]: OK

**Review by tmysik — APPROVED** [review]

**Comment by tmysik** [comment]
@junichi11 Perfectly done, thank you!

**Comment by tmysik** [comment]
> > According to RFC name of argument can be also reserved keyword.
> 
> Yes, I already noticed that. As you wrote, it's not a big problem. Maybe, it's not easy. I'm going to try to fix it later.

I guess this will be difficult to fix.

**Comment by tmysik** [comment]
> Note: other features are not fixed yet (e.g. Code completion). I'll fix them later.

Code completion would be really nice!

**Comment by tmysik** [comment]
@KacerCZ Thank you for your review!

@junichi11 As I can see, you don't need me guys :)

**Comment by junichi11** [comment]
@KacerCZ @tmysik 
Thank you for your reviews :)

@tmysik
No way! It's my pleasure to receive your review. I can merge without your review if you can't make time for it. However, I would feel sad without you because I really like working on something with you... (and of course, I really like your code :))

**Comment by junichi11** [comment]
> > > According to RFC name of argument can be also reserved keyword.
> > 
> > 
> > Yes, I already noticed that. As you wrote, it's not a big problem. Maybe, it's not easy. I'm going to try to fix it later.
> 
> I guess this will be difficult to fix.

If my idea is correct, I can fix this ;)

nb-php80-named-arguments-reserved-keyword-highlighting

**Comment by tmysik** [comment]
@junichi11 It is and always was a pleasure to work with you!

**Pull request merged**